asm/instruction.go

```diff
@@ -127,30 +131,53 @@ func (ins Instruction) Marshal(w io.Writer, bo binary.ByteOrder) (uint64, error)
 	return 2 * InstructionSize, nil
 }
 
-// RewriteMapPtr changes an instruction to use a new map fd.
+// AssociateMap associates a *ebpf.Map with this instruction.
 //
 // Returns an error if the instruction doesn't load a map.
-func (ins *Instruction) RewriteMapPtr(fd int) error {
-	if !ins.OpCode.IsDWordLoad() {
-		return fmt.Errorf("%s is not a 64 bit load", ins.OpCode)
+func (ins *Instruction) AssociateMap(m FDer) error {
+	if !ins.IsLoadFromMap() {
+		return errors.New("not a load from a map")
 	}
 
-	if ins.Src != PseudoMapFD && ins.Src != PseudoMapValue {
+	// Set fd to -1 since it is technically not yet rewritten
+	ins.encodeMapFD(-1)
+	ins.setMap(m)
+	return nil
+}
+
+// RewriteMapPtr changes an instruction to use a new map fd.
+//
+// Returns an error if the instruction doesn't load a map.
+//
+// Deprecated: use AssociateMap instead.
+func (ins *Instruction) RewriteMapPtr(fd int) error {
+	if !ins.IsLoadFromMap() {
 		return errors.New("not a load from a map")
 	}
 
+	ins.encodeMapFD(fd)
+
+	return nil
+}
+
+func (ins *Instruction) encodeMapFD(fd int) {
 	// Preserve the offset value for direct map loads.
 	offset := uint64(ins.Constant) & (math.MaxUint32 << 32)
 	rawFd := uint64(uint32(fd))
 	ins.Constant = int64(offset | rawFd)
-	return nil
 }
 
 // MapPtr returns the map fd for this instruction.
 //
 // The result is undefined if the instruction is not a load from a map,
 // see IsLoadFromMap.
 func (ins *Instruction) MapPtr() int {
```

Comment:
Don't know if we want to just yet, I recall an earlier change to this PR suggested by @ti-mo to still support use cases where users can specify their own FD's instead of using our facilities. Though I suppose a user could always implement their own `FDer` to cover that scenario. 

Perhaps we should then also provide a `Map() FDer` getter? Or decide to not support custom FD setting from outside the lib.

Reply:
Maybe the comment was a bit terse, this method should be sunset with `// Deprecated:` to be removed in 0.9.

Caller-supplied FDer would indeed be an excellent workaround for users that are somehow unable to switch. This should go into the patch notes and in the deprecation notices on both `MapPtr()` and `RewriteMapPtr()`.

Reply:
Makes me wish I had used `Map.Fd` instead of `Map.FD`. That way callers could use `os.File` as an Fder.